### PR TITLE
Ignore permalink_prefix when serializing pills

### PR DESCRIPTION
**Pull request by herkulessi**
## Checklist

-   [X] Tests written for new code (and old code if feasible)
-   [X] Linter and other CI checks pass
-   [X] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md))

fixes vector-im/element-web/issues/26002

During serialization of messages, pills were wrongfully serialized to a URL
starting with `permalink_prefix`. This is against the Spec (which mandates
https://matrix.to/#/ links) and the resulting pills were not recognized as
pills in other clients.

Spec-Appendix concerning matrix.to links: https://spec.matrix.org/v1.8/appendices/#matrixto-navigation

Signed-off-by: Lars Wickel <git@herkulessi.de>

---
This PR currently has none of the required changelog labels.

A reviewer can add one of: `T-Deprecation`, `T-Enhancement`, `T-Defect`, `T-Task` to indicate what type of change this is, or add `Type: [enhancement/defect/task]` to the description and I'll add them for you.

### CODE REVIEW

File: src/utils/permalinks/ElementPermalinkConstructor.ts [hunk 1]
```diff
@@ -31,23 +31,32 @@ export default class ElementPermalinkConstructor extends PermalinkConstructor {
         }
     }
 
-    public forEvent(roomId: string, eventId: string, serverCandidates: string[]): string {
+    public forEvent(roomId: string, eventId: string, serverCandidates: string[], ispill = false): string {
+        if (ispill) {
+            return `https://matrix.to/#/${roomId}/${eventId}${this.encodeServerCandidates(serverCandidates)}`;
```
Comment (dbkr) [inline src/utils/permalinks/ElementPermalinkConstructor.ts:36]: This syntax is already generated by MatrixToPermalinkConstructor. I think probably the right fix here would be to pass the flag into `makeGenericPermalink` but then have `getPermalinkConstructor()` return the matrix.to one instead. This should make this change a lot simpler as all the various methods of the different PermalinkConstructor classes won't need the extra param.

File: src/utils/permalinks/Permalinks.ts [hunk 2]
```diff
@@ -269,26 +269,26 @@ export class RoomPermalinkCreator {
     };
 }
 
-export function makeGenericPermalink(entityId: string): string {
-    return getPermalinkConstructor().forEntity(entityId);
+export function makeGenericPermalink(entityId: string, ispill = false): string {
```
Comment (dbkr) [inline src/utils/permalinks/Permalinks.ts:272]: Could the param follow the code style & be camelCased to isPill please? I think I read it asm, "I spill" the first time, and that's in the context of this PR.

Also this function could really benefit from some doc now it's not just a simple single-parameter function, as it's not really obvious how the isPill parameter changes the function's behaviour.

**Review by dbkr — CHANGES_REQUESTED** [review]

File: src/utils/permalinks/Permalinks.ts (hunk 2 above)
Comment (herkulessi) [reply src/utils/permalinks/Permalinks.ts:272]: Where/How should I document that? A separate file in docs/ seems a bit overkill...
Just as comments in the source?

Comment (dbkr) [reply src/utils/permalinks/Permalinks.ts:272]: Yeah, a jsdoc comment is perfect (eg. https://github.com/matrix-org/matrix-react-sdk/blob/a3240242e8f581512570badd8c39f57eb124d540/src/utils/permalinks/Permalinks.ts#L305)

**Review by dbkr — APPROVED** [review]
Thanks, sorry for missing the update on this. (Also, please don't force-push PRs: it breaks all GitHub's review features).

**Pull request merged**